cranelift-codegen/src/isa/x86/unwind.rs

```diff
@@ -0,0 +1,508 @@
+enum UnwindCode {
+    PushRegister { offset: u8, reg: RegUnit },
+    StackAlloc { offset: u8, size: u32 },
+    SetFramePointer { offset: u8, sp_offset: u8 },
+}
+
+impl UnwindCode {
+    fn emit(&self, mem: &mut Vec<u8>) -> std::io::Result<()> {
+        enum UnwindOperation {
+            PushNonvolatileRegister,
+            LargeStackAlloc,
+            SmallStackAlloc,
+            SetFramePointer,
+        }
+
+        match self {
+            Self::PushRegister { offset, reg } => {
+                mem.write_u8(*offset)?;
+                mem.write_u8(
+                    ((*reg as u8) << 4) | (UnwindOperation::PushNonvolatileRegister as u8),
+                )?;
+            }
+            Self::StackAlloc { offset, size } => {
+                // Stack allocations on Windows must be a multiple of 8 and be at least 1 slot
+                assert!(*size >= 8);
+                assert!((*size % 8) == 0);
+
+                mem.write_u8(*offset)?;
+                if *size <= SMALL_ALLOC_MAX_SIZE {
+                    mem.write_u8(
+                        ((((*size - 8) / 8) as u8) << 4) | UnwindOperation::SmallStackAlloc as u8,
+                    )?;
+                } else if *size <= LARGE_ALLOC_16BIT_MAX_SIZE {
+                    mem.write_u8(UnwindOperation::LargeStackAlloc as u8)?;
+                    mem.write_u16::<LittleEndian>((*size / 8) as u16)?;
+                } else {
+                    mem.write_u8((1 << 4) | (UnwindOperation::LargeStackAlloc as u8))?;
+                    mem.write_u32::<LittleEndian>(*size)?;
+                }
+            }
+            Self::SetFramePointer { offset, sp_offset } => {
+                mem.write_u8(*offset)?;
+                mem.write_u8((*sp_offset << 4) | (UnwindOperation::SetFramePointer as u8))?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn node_count(&self) -> usize {
+        match self {
+            Self::StackAlloc { size, .. } => {
+                if *size <= SMALL_ALLOC_MAX_SIZE {
+                    1
+                } else if *size <= LARGE_ALLOC_16BIT_MAX_SIZE {
+                    2
+                } else {
+                    3
+                }
+            }
+            _ => 1,
+        }
+    }
+}
+
+/// Represents Windows x64 unwind information.
+///
+/// For information about Windows x64 unwind info, see:
+/// https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64
+#[derive(Debug, PartialEq, Eq)]
+pub struct UnwindInfo {
+    flags: u8,
+    prologue_size: u8,
+    frame_register: Option<RegUnit>,
+    frame_register_offset: u8,
+    unwind_codes: Vec<UnwindCode>,
+}
+
+impl UnwindInfo {
+    pub fn try_from_func(
+        func: &Function,
+        isa: &dyn TargetIsa,
+        frame_register: Option<RegUnit>,
+    ) -> Option<Self> {
+        // Only Windows fastcall is supported for unwind information
+        if func.signature.call_conv != CallConv::WindowsFastcall || func.prologue_end.is_none() {
```

Comment:
In functions that don't require their own stack space, we don't currently emit a prologue, so `prologue_end` will be `None` here. Is it intended to emit no unwind info for such functions?

Reply:
In Windows x64 ABI, the only functions that don't need unwind information are "leaf functions", which have specific requirements:

* cannot modify non-volatile registers
* cannot call other functions
* does not need to re-align the stack (i.e. don't need a prologue).
* can be unwound via simulation of a return instruction

As such, they do not require unwind information because the OS can simply "simulate a return" for any function that doesn't have unwind information.

However, frame functions (i.e. non-leaf functions) must have prologues because they are required to realign the stack even if they have no need for their own stack space, provided the frame function calls another function.  If the frame function does not call another function, then it is not required to realign the stack, but then there's probably another reason why the function isn't a "leaf" and would require a prologue anyway (stack allocation, saving of non-volatile registers, etc).

Reply:
Looking over the code, it does seem like we currently always generate a function prologue for fastcall functions that, at least, allocates homing space for callees (even if the function doesn't call), so we could instead panic if the prologue end marker isn't set and unwind information is requested, rather than silently omit unwind information.

In the future, the code generation should probably omit frame pointers and detect "leaf" functions that don't need prologues.  At that point we could stop panicking and instead not emit unwind information.

Reply:
FYI: I've filed https://github.com/CraneStation/cranelift/issues/1201 to track potential improvements for fastcall code generation.